### PR TITLE
Add !online command

### PR DESCRIPTION
**Pull request by laurirasanen**
image

### CODE REVIEW

File: format.js [hunk 1]
```diff
@@ -61,6 +61,29 @@ function format_map(m) {
     return embed;
 }
 
+function format_online(players) {
+    players.sort((a, b) => {
+        if (a.rank_info.rank < b.rank_info.rank) return -1;
```
Comment (arispoloway) [inline format.js]: You could just subtract them :)

File: handlers.js [hunk 2]
```diff
@@ -99,6 +115,23 @@ let [handle_stime, handle_dtime] = both_classes(time);
...
 let [handle_srank, handle_drank] = both_classes(rank);
 
+async function handle_online() {
+    let servers = await tempus.serverList();
+    let players = [];
+    for (let s of servers) {
+        if (!s.game_info) continue;
+        for (let p of s.game_info.players) {
+            if (!p.id || p.id == null) continue;
+            let player = await p.toPlayerStats();
```
Comment (arispoloway) [inline handlers.js]: Would it be possible to do this in parallel via some sort of promise map or something? Seems rather annoying to have to do this all sequentially.

File: handlers.js [hunk 3]
```diff
@@ -79,6 +79,22 @@ async function rank(c, player, num) {
     }
 }
 
+function reply_wait(f)
+{
```
Comment (arispoloway) [inline handlers.js]: Can you put this on the same line?

File: handlers.js [hunk 4]
```diff
@@ -79,6 +79,22 @@ async function rank(c, player, num) {
...
+    return async function slow_function(reply, args)
+    {
```
Comment (arispoloway) [inline handlers.js]: Same here

File: handlers.js [hunk 5]
```diff
@@ -79,6 +79,22 @@ async function rank(c, player, num) {
     }
 }
 
+function reply_wait(f)
```
Comment (arispoloway) [inline handlers.js]: I like the idea of this function - I do think it would be nice to be able to edit the message instead though. It looks like I've definitely over-abstracted some of this stuff and it doesn't fit so nicely with messages like this. If you can think of a nice way to add that functionality without changing too much that would be great, but if not its no big deal - this is fine!

File: handlers.js [hunk 6]
```diff
@@ -155,7 +188,11 @@ async function handle_message(reply, content) {
     handler = handlers[content.split(" ")[0]];
     if (handler != undefined) {
         try{
-            let r = await handler(content.split(" ").slice(1, 9));
+            var args = [content.split(" ").slice(1, 9)];
```
Comment (arispoloway) [inline handlers.js]: Why not leave out the extra enclosing `[]` and refer to it differently below?

File: handlers.js [hunk 7]
```diff
@@ -155,7 +188,11 @@ async function handle_message(reply, content) {
...
+            if (handler.name == reply_wait().name) {
```
Comment (arispoloway) [inline handlers.js]: This feels a little funky to me - is there a cleaner way to approach this? I don't like the idea of hardcoding for this specific handler here if it can be avoided. Also, can you not just use `reply_wait.name` here?

**Review by arispoloway — COMMENTED** [review]
Do you have a rough idea of how long this might expect to take on average? Overall nice work - seems to be a bit awkward in places but I'm sure lots of that is due to my own overfitting.

File: format.js (hunk 1 above)
Comment (laurirasanen) [reply format.js]: Resolved in https://github.com/arispoloway/tempus-discord-bot/pull/2/commits/63778ca1eccd7944977fd2ed403f5471db2dd379

File: handlers.js (hunk 3 above)
Comment (laurirasanen) [reply handlers.js]: Resolved in https://github.com/arispoloway/tempus-discord-bot/pull/2/commits/63778ca1eccd7944977fd2ed403f5471db2dd379

File: handlers.js (hunk 4 above)
Comment (laurirasanen) [reply handlers.js]: Resolved in https://github.com/arispoloway/tempus-discord-bot/pull/2/commits/63778ca1eccd7944977fd2ed403f5471db2dd379

File: handlers.js (hunk 7 above)
Comment (laurirasanen) [reply handlers.js]: It is definitely funky, the first alternative that comes to mind would be to just pass the message context, or the reply function, to all the handlers even if they don't use it. `reply_wait.name` is just going to return "reply_wait". Adding the parentheses will return the enclosed `slow_function`.

File: handlers.js (hunk 6 above)
Comment (laurirasanen) [reply handlers.js]: The alternative would result in more if/else statements unless I'm missing something. Although I think even this looks cleaner.
```js
let r;
if (handler.name == reply_wait().name) {
    r = await handler(reply, content.split(" ").slice(1, 9));
}
else {
    r = await handler(content.split(" ").slice(1, 9));
}
```

**Comment by laurirasanen** [comment]
Takes about 20-25 seconds to respond with a list of players right now. I'll look at parallelization tomorrow.

**Comment by laurirasanen** [comment]
https://github.com/arispoloway/tempus-discord-bot/pull/2/commits/89778b3b14169856dc46e03ee1bc5844345a5a6e seems to drop this down to 7-8 seconds.

File: handlers.js (hunk 2 above)
Comment (laurirasanen) [reply handlers.js]: Resolved in https://github.com/arispoloway/tempus-discord-bot/commit/89778b3b14169856dc46e03ee1bc5844345a5a6e

File: bot.js [hunk 8]
```diff
@@ -13,15 +13,15 @@ update_interval = 1000 * 60 * 60 * 6 // every 6 hours
...
+    return (reply, previous = null) => {
         if (settings.logging) {
             fs.appendFile(path.resolve(__dirname, settings.logging), 
                 `${new Date().getTime()}` + 
                 `|${msg.author.username}|${msg.author.id}|${msg.content}|${msg.guild ? msg.guild.name : undefined}|${msg.channel.name}\n`,
                 () => {});
         }
 
-        return msg.channel.send(reply);
+        return previous ? previous.edit(reply) : msg.channel.send(reply);
```
Comment (laurirasanen) [inline bot.js]: Might need to check for `previous.editable` here for future-proofing, I think you can always edit your own messages as of right now.

File: handlers.js [hunk 9]
```diff
@@ -191,17 +185,11 @@ async function handle_message(reply, content) {
...
     if (handler != undefined) {
         try{
-            let r;
-            if (handler.name == reply_wait().name) {
-                r = await handler(reply, content.split(" ").slice(1, 9));
-            }
-            else {
-                r = await handler(content.split(" ").slice(1, 9));
-            }
+            let r = await handler(content.split(" ").slice(1, 9), reply);
```
Comment (laurirasanen) [inline handlers.js:186]: Passing `reply` to all the handlers should be fine, it just won't get used, though I imagine there's a nicer way to do this.

File: handlers.js [hunk 10]
```diff
@@ -191,17 +185,11 @@ async function handle_message(reply, content) {
...
             if (!r) {
                 reply(format.format_error("Invalid arguments"));
             } else {
-                reply(r);
+                reply(...typeof r[Symbol.iterator] == "function" ? r : [r]);
```
Comment (laurirasanen) [inline handlers.js:192]: Not a huge fan of this check, not sure how else to handle `reply_wait` returning multiple values.

File: bot.js (hunk 8 above)
Comment (laurirasanen) [reply bot.js]: Resolved in https://github.com/arispoloway/tempus-discord-bot/pull/2/commits/1eaf4d7c8116bb84f7a1acaf348c521b68f76ffe

File: handlers.js [hunk 11]
```diff
@@ -99,6 +107,28 @@ let [handle_stime, handle_dtime] = both_classes(time);
...
+            .map(p => p.toPlayerStats().then(player => {
+                    player.server = s;
+                    return player;
+                })
+            )
+        )
+        .reduce((prev, next) => prev.concat(next));
+
+    return await Promise.all(player_promises)
+        .then((players) => {
```
Comment (arispoloway) [inline handlers.js]: Can you convert this to await form?

File: handlers.js [hunk 12]
```diff
@@ -99,6 +107,28 @@ let [handle_stime, handle_dtime] = both_classes(time);
 let [handle_sdem, handle_ddem] = both_classes(dem);
 let [handle_srank, handle_drank] = both_classes(rank);
 
+async function handle_online() {
+    let servers = await tempus.serverList();
+    let player_promises = servers
```
Comment (arispoloway) [inline handlers.js:112]: Beautiful

File: handlers.js (hunk 9 above)
Comment (arispoloway) [reply handlers.js:186]: Seems fine to me :+1:

File: bot.js [hunk 13]
```diff
@@ -13,15 +13,19 @@ update_interval = 1000 * 60 * 60 * 6 // every 6 hours
...
         if (settings.logging) {
             fs.appendFile(path.resolve(__dirname, settings.logging), 
                 `${new Date().getTime()}` + 
                 `|${msg.author.username}|${msg.author.id}|${msg.content}|${msg.guild ? msg.guild.name : undefined}|${msg.channel.name}\n`,
                 () => {});
         }
+
+        if (previous && !previous.deleted) { 
+            if (previous.editable) return previous.edit(reply);
+            if (previous.deletable) return msg.channel.send.reply().then(previous.delete());
```
Comment (arispoloway) [inline bot.js]: Does `previous.delete()` return a promise? Regardless, can you use await format here? I'd prefer to keep things as consistent as possible.

File: handlers.js (hunk 10 above)
Comment (arispoloway) [reply handlers.js:192]: Yeah it's definitely a bit strange but something like this is needed I guess. Can you just put in a comment to explain why you're doing this in case someone tries to come back and understand? It's a bit hard to parse without proper context.

**Review by arispoloway — COMMENTED** [review]
Looks pretty good just a few minor things now. Probably some stuff I could consider refactoring but definitely not as part of this.
